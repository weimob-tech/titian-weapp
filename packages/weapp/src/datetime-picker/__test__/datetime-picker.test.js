const simulate = require('miniprogram-simulate');
const { getComponent, getComponentId } = require('../../common/test/getComponent');

const mockToday = '2022-02-27';
jest.setSystemTime(new Date(mockToday));
const { data } = require('./mock/index');

beforeAll(() => {
  jest.useFakeTimers('modern');
});
describe('test component', () => {
  it('should render success', () => {
    const comp = getComponent('datetime-picker');
    expect(comp).toMatchSnapshot();
  });

  it('should render success', async () => {
    const { props } = data[6];
    const datetimePicker = getComponentId('datetime-picker');
    const changeCallback = jest.fn();

    const propStr = Object.keys(props).reduce((target, key) => {
      const lineKey = key.replace(/([a-z])([A-Z])/, '$1-$2').toLowerCase();
      return ` ${target}   ${lineKey}="{{ ${key} }}" `;
    }, '');
    const wrapper = simulate.render(
      simulate.load({
        usingComponents: {
          'ti-datetime-picker': datetimePicker
        },
        data: props,
        template: `<ti-datetime-picker
                        id="datetimePicker"
                        ${propStr}
                        bind:change="changeCallback"   /> `,
        methods: {
          changeCallback
        }
      })
    );
    wrapper.attach(document.createElement('parent-datetime-picker'));
    await Promise.resolve();
    changeCallback.mockClear();
    const yearComp = wrapper
      .querySelectorAll('#datetimePicker >>> .internal_children')
      .find((item) => item.data.colAlias === 'year');
    yearComp.instance.onSelect({
      rowIndex: yearComp.data.row + 1,
      colAlias: yearComp.data.colAlias
    });
    await Promise.resolve();
    expect(changeCallback).toHaveBeenCalledTimes(1);
  });

  test.each(data)(`options $label `, async (item) => {
    const datetimePicker = getComponentId('datetime-picker');
    const changeCallback = jest.fn();
    const toggleVisible = jest.fn();

    const propStr = Object.keys(item.props).reduce((target, key) => {
      const lineKey = key.replace(/([a-z])([A-Z])/, '$1-$2').toLowerCase();
      return ` ${target}   ${lineKey}="{{ ${key} }}" `;
    }, '');
    const wrapper = simulate.render(
      simulate.load({
        usingComponents: {
          'ti-datetime-picker': datetimePicker
        },
        data: item.props,
        template: `<ti-datetime-picker
                        id="datetimePicker"
                        ${propStr}
                        bind:change="changeCallback" bind:cancel="toggleVisible" bind:confirm="toggleVisible" /> `,
        methods: {
          changeCallback,
          toggleVisible
        }
      })
    );
    wrapper.attach(document.createElement('parent-datetime-picker'));
    await Promise.resolve();
    expect(changeCallback).toHaveBeenCalledTimes(item.expect[0].number);
    expect(changeCallback.mock.calls[0][0].detail).toStrictEqual(item.expect[0].data);
    if (typeof item.props.formatter === 'function') {
      const { colAlias } = wrapper.querySelector('#datetimePicker >>> .titian-picker-column-item-box').dom.dataset;
      const dom = wrapper.querySelector('#datetimePicker >>> .titian-picker-column-item').dom.innerHTML.trim();
      const val = wrapper.querySelector('#datetimePicker').data[`${colAlias}Option`][0].value;
      item.props.formatter(colAlias, val);
      expect(item.props.formatter(colAlias, val)).toBe(dom);
    }

    if (item.expect[0].first) {
      const firstColumn = wrapper.querySelector('#datetimePicker >>> .titian-picker-column-item-box');
      expect(firstColumn.dom.dataset.colAlias).toBe(item.expect[0].first);
    }

    const confirm = wrapper.querySelector('#datetimePicker >>> .titian-titlebar-right');
    confirm.dispatchEvent('tap');
    await Promise.resolve();
    expect(toggleVisible).toHaveBeenCalledTimes(1);
    expect(toggleVisible.mock.calls[0][0].detail).toStrictEqual(item.expect[0].data);

    toggleVisible.mockClear();

    const cancel = wrapper.querySelector('#datetimePicker >>> .titian-titlebar-left');
    cancel.dispatchEvent('tap');
    await Promise.resolve();
    expect(toggleVisible).toHaveBeenCalledTimes(1);
    toggleVisible.mockClear();
  });
});
afterAll(() => {
  jest.useRealTimers();
});
