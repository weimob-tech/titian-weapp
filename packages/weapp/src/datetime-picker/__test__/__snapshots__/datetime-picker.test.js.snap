// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test component should render success 1`] = `
<main>
  <ti-picker
    class="titian-datetime-picker"
    extClass="ext-class"
    extHairlineClass="ext-hairline-class"
    extMaskClass="ext-mask-class"
    catch:cancel="onCalendarCancel"
    catch:change="onCalendarChange"
    catch:confirm="onCalendarConfirm"
    catch:reach-bottom="noop"
    catch:reach-top="noop"
  >
    <wx-view
      class="titian-picker"
    >
      <ti-popup-titlebar
        onlyCancel="{{true}}"
        catch:cancel="onCancel"
        catch:confirm="onConfirm"
      >
        <ti-title-bar
          extClass="ext-class"
          extLeftClass="ext-left-class titian-popup-titlebar-cancel"
          extRightClass="ext-right-class titian-popup-titlebar-confirm"
          extRightIconClass="ext-right-icon-class "
          extTitleClass="ext-title-class "
          bind:click="onClick"
        >
          <wx-view
            class="ext-class titian-titlebar"
            style=""
          >
            <wx-view
              class="ext-left-class titian-titlebar-left"
              bind:tap="onClickLeft"
            />
            <wx-view
              class="ext-title-class titian-titlebar-title"
            >
              <wx-view
                class="ext-main-title-class titian-titlebar-main-title"
                bind:tap="onClickTitle"
              >
                
          标题
        
              </wx-view>
            </wx-view>
            <wx-view
              class="ext-right-class"
              bind:tap="onClickRight"
            />
            <wx-view
              class="ext-right-class titian-titlebar-right"
              bind:tap="onClickRight"
            />
            <wx-view
              class="ext-left-class titian-titlebar-left"
              bind:tap="onClickLeft"
            />
            <wx-view
              class="ext-title-class titian-titlebar-title"
            >
              <wx-view
                class="ext-main-title-class titian-titlebar-main-title"
                bind:tap="onClickTitle"
              >
                
          标题
        
              </wx-view>
            </wx-view>
            <wx-view
              class="ext-right-class"
              bind:tap="onClickRight"
            />
            <wx-view
              class="ext-right-class titian-titlebar-right"
              bind:tap="onClickRight"
            />
          </wx-view>
        </ti-title-bar>
      </ti-popup-titlebar>
      <wx-view
        class="titian-picker-content ext-class"
        disableScroll="{{true}}"
        style=";height:0px"
      >
        <wx-view
          class="titian-picker-loading"
        >
          <ti-loading>
            <wx-view
              class="titian-loading-wrap"
              style=""
            >
              <wx-view
                class="ext-class titian-loading-circular titian-loading"
                style="padding:6rpx;width:60rpx;height:60rpx"
              >
                <wx-view
                  class="titian-loading-view-circular titian-loading-view"
                  style="width:60rpx;height:60rpx"
                >
                  <wx-view
                    class="titian-loading-view-item"
                    style="width:60rpx;height:60rpx;border-width:5rpx"
                  />
                </wx-view>
                <wx-view
                  class="titian-loading-view-circular titian-loading-view"
                  style="width:60rpx;height:60rpx"
                >
                  <wx-view
                    class="titian-loading-view-item"
                    style="width:60rpx;height:60rpx;border-width:5rpx"
                  />
                </wx-view>
                <wx-view
                  class="titian-loading-view-circular titian-loading-view"
                  style="width:60rpx;height:60rpx"
                >
                  <wx-view
                    class="titian-loading-view-item"
                    style="width:60rpx;height:60rpx;border-width:5rpx"
                  />
                </wx-view>
                <wx-view
                  class="titian-loading-view-circular titian-loading-view"
                  style="width:60rpx;height:60rpx"
                >
                  <wx-view
                    class="titian-loading-view-item"
                    style="width:60rpx;height:60rpx;border-width:5rpx"
                  />
                </wx-view>
                <wx-view
                  class="titian-loading-view-circular titian-loading-view"
                  style="width:60rpx;height:60rpx"
                >
                  <wx-view
                    class="titian-loading-view-item"
                    style="width:60rpx;height:60rpx;border-width:5rpx"
                  />
                </wx-view>
              </wx-view>
              <wx-view
                class="titian-loading-text"
                style=""
              />
            </wx-view>
          </ti-loading>
        </wx-view>
        <wx-view
          class="titian-picker-mask ext-mask-class"
        />
        <wx-view
          class="titian-picker-hairline ext-hairline-class"
          style=" height:0px;top : 0px"
        />
        <wx-view
          class="titian-picker-column-box"
        />
      </wx-view>
    </wx-view>
  </ti-picker>
</main>
`;
